File: JavaScript/Interview_problems/Payment_Service_Demo.js
/*
================================================================================
LLD: Payment Service
================================================================================

Functional Requirements:
    1. User should be able to create a payment.
    2. Support multiple payment methods:
       - Credit Card
       - Debit Card
       - UPI
       - Net Banking
       - Wallet
    3. Prevent duplicate payments using an idempotency key.
    4. Retry failed payments.
    5. Payment Gateway should be pluggable.

Non-Functional Requirements:
    - Service should avoid duplicate payments.
    - Security (basic input validation before a request ever reaches a
      gateway).
    - Extensible.
    - Thread safe (Node's event loop makes this moot for the sync path
      below, but the design would carry over to a worker-thread model).
    - Retry support.

Design:
    PaymentMethod (Strategy) validates a request is well-formed for its
    channel (card number, UPI id, bank code, ...) before any money moves.

    PaymentGateway (Strategy) is the pluggable boundary to an external
    processor (RazorPay, Stripe, ...). PaymentGatewayDecorator wraps a
    PaymentGateway with the same interface, and RetryLogic is a concrete
    decorator that re-attempts a failed call up to N times -- this is what
    the original stub only sketched: it called the inner gateway, checked
    for failure, and then did nothing, so a failed payment just vanished
    with no retry and no response.

    IdempotencyManager is a key -> PaymentResponse cache. A repeat request
    with a key already seen returns the cached result instead of charging
    the user twice -- this is what the original stub got wrong: it checked
    "not a duplicate" before submitting the work, but never checked the
    "is a duplicate" branch it also declared, and never records anything
    into the manager, so every repeat request would have reprocessed the
    payment.

    PaymentService wires a pluggable gateway, per-type payment methods, and
    the idempotency cache together, and awaits the gateway's result so
    createPayment can actually return the outcome to the caller -- unlike
    the original stub, which fired the gateway call and threw the result
    away.

Core Entities:
    PaymentRequest / PaymentResponse / PaymentType / PaymentStatus
    PaymentMethod (Card / UPI / NetBanking)
    PaymentGateway (RazorPay / Stripe) / PaymentGatewayDecorator / RetryLogic
    IdempotencyManager
    PaymentService
================================================================================
*/

const PaymentType = Object.freeze({
    CARD: "CARD",
    UPI: "UPI",
    NET_BANKING: "NET_BANKING",
    WALLET: "WALLET",
});

const PaymentStatus = Object.freeze({
    SUCCESS: "SUCCESS",
    FAILED: "FAILED",
    DUPLICATE: "DUPLICATE",
});

class PaymentRequest {
    constructor(idempotencyKey, userId, amount, type, details) {
        this.idempotencyKey = idempotencyKey;
        this.userId = userId;
        this.amount = amount;
        this.type = type;
        this.details = details;
    }
}

class PaymentResponse {
    constructor(paymentId, status, message) {
        this.paymentId = paymentId;
        this.status = status;
        this.message = message;
    }

    toString() {
        return `PaymentResponse(payment_id=${this.paymentId ?? "null"}, status=${this.status}, message=${this.message})`;
    }
}

class PaymentMethod {
    validate(request) {
        throw new Error("validate() must be implemented");
    }
}

class CardPaymentMethod extends PaymentMethod {
    validate(request) {
        const cardNumber = request.details.cardNumber;
        return typeof cardNumber === "string" && cardNumber.length === 16;
    }
}

class UpiPaymentMethod extends PaymentMethod {
    validate(request) {
        const upiId = request.details.upiId;
        return typeof upiId === "string" && upiId.includes("@");
    }
}

class NetBankingPaymentMethod extends PaymentMethod {
    validate(request) {
        const bankCode = request.details.bankCode;
        return typeof bankCode === "string" && bankCode.length > 0;
    }
}

class PaymentGateway {
    async process(request) {
        throw new Error("process() must be implemented");
    }
}

// Simulates a gateway that fails the first attempt for a given idempotency
// key and succeeds on every attempt after that -- enough to exercise
// RetryLogic deterministically, without relying on real network flakiness.
class RazorPayGateway extends PaymentGateway {
    constructor() {
        super();
        this._attempts = new Map();
    }

    async process(request) {
        const attempt = (this._attempts.get(request.idempotencyKey) || 0) + 1;
        this._attempts.set(request.idempotencyKey, attempt);
        if (attempt === 1) {
            return new PaymentResponse(null, PaymentStatus.FAILED, "RazorPay: transient gateway error (attempt 1)");
        }
        return new PaymentResponse(`razorpay-${request.idempotencyKey}`, PaymentStatus.SUCCESS,
            `RazorPay: payment captured (attempt ${attempt})`);
    }
}

// A second, always-succeeding gateway -- demonstrates that PaymentService
// only depends on the PaymentGateway interface, so the processor behind it
// is pluggable.
class StripeGateway extends PaymentGateway {
    async process(request) {
        return new PaymentResponse(`stripe-${request.idempotencyKey}`, PaymentStatus.SUCCESS, "Stripe: payment captured");
    }
}

class PaymentGatewayDecorator extends PaymentGateway {
    constructor(gateway) {
        super();
        this.gateway = gateway;
    }
}

class RetryLogic extends PaymentGatewayDecorator {
    constructor(gateway, maxAttempts) {
        super(gateway);
        this.maxAttempts = maxAttempts;
    }

    async process(request) {
        let lastResponse = null;
        for (let attempt = 1; attempt <= this.maxAttempts; attempt++) {
            lastResponse = await this.gateway.process(request);
            if (lastResponse.status === PaymentStatus.SUCCESS) {
                return lastResponse;
            }
        }
        return lastResponse;
    }
}

// key -> the response that was returned the first time that key was seen.
class IdempotencyManager {
    constructor() {
        this._cache = new Map();
    }

    getCached(idempotencyKey) {
        return this._cache.get(idempotencyKey) ?? null;
    }

    record(idempotencyKey, response) {
        this._cache.set(idempotencyKey, response);
    }
}

class PaymentService {
    constructor(gateway) {
        this._methods = {
            [PaymentType.CARD]: new CardPaymentMethod(),
            [PaymentType.UPI]: new UpiPaymentMethod(),
            [PaymentType.NET_BANKING]: new NetBankingPaymentMethod(),
        };
        this._gateway = gateway;
        this._idempotencyManager = new IdempotencyManager();
    }

    async createPayment(request) {
        const cached = this._idempotencyManager.getCached(request.idempotencyKey);
        if (cached !== null) {
            return new PaymentResponse(cached.paymentId, PaymentStatus.DUPLICATE,
                `Duplicate request for idempotency key ${request.idempotencyKey}; returning cached result`);
        }

        const method = this._methods[request.type];
        if (!method || !method.validate(request)) {
            return new PaymentResponse(null, PaymentStatus.FAILED, `Validation failed for ${request.type}`);
        }

        const response = await this._gateway.process(request);
        this._idempotencyManager.record(request.idempotencyKey, response);
        return response;
    }
}

async function main() {
    const service = new PaymentService(new RetryLogic(new RazorPayGateway(), 3));

    console.log("1) Valid card payment (gateway fails attempt #1, RetryLogic recovers it):");
    const r1 = await service.createPayment(new PaymentRequest("key-1", "user-1", 250.0, PaymentType.CARD,
        { cardNumber: "4111111111111111" }));
    console.log(`   ${r1}`);

    console.log("\n2) Same idempotency key resubmitted (must not charge again):");
    const r2 = await service.createPayment(new PaymentRequest("key-1", "user-1", 250.0, PaymentType.CARD,
        { cardNumber: "4111111111111111" }));
    console.log(`   ${r2}`);

    console.log("\n3) Invalid UPI id (missing '@'):");
    const r3 = await service.createPayment(new PaymentRequest("key-2", "user-2", 75.0, PaymentType.UPI,
        { upiId: "not-a-valid-upi-id" }));
    console.log(`   ${r3}`);

    console.log("\n4) Valid UPI payment on a new idempotency key (also recovers via retry):");
    const r4 = await service.createPayment(new PaymentRequest("key-3", "user-2", 75.0, PaymentType.UPI,
        { upiId: "user2@upi" }));
    console.log(`   ${r4}`);

    console.log("\n5) Swapping in Stripe for a net-banking payment (gateway is pluggable):");
    const stripeService = new PaymentService(new RetryLogic(new StripeGateway(), 3));
    const r5 = await stripeService.createPayment(new PaymentRequest("key-4", "user-3", 500.0, PaymentType.NET_BANKING,
        { bankCode: "HDFC0001" }));
    console.log(`   ${r5}`);
}

main();

File: Golang/Interview_problems/Payment_Service_Demo.go
/*
================================================================================
LLD: Payment Service
================================================================================

Functional Requirements:
    1. User should be able to create a payment.
    2. Support multiple payment methods:
       - Credit Card
       - Debit Card
       - UPI
       - Net Banking
       - Wallet
    3. Prevent duplicate payments using an idempotency key.
    4. Retry failed payments.
    5. Payment Gateway should be pluggable.

Non-Functional Requirements:
    - Service should avoid duplicate payments.
    - Security (basic input validation before a request ever reaches a
      gateway).
    - Extensible.
    - Thread safe.
    - Retry support.

Design:
    PaymentMethod (Strategy) validates a request is well-formed for its
    channel (card number, UPI id, bank code, ...) before any money moves.

    PaymentGateway (Strategy) is the pluggable boundary to an external
    processor (RazorPay, Stripe, ...). RetryLogic wraps a PaymentGateway
    with the same interface and re-attempts a failed call up to N times --
    this is what the original stub only sketched: it called the inner
    gateway, checked for failure, and then did nothing, so a failed payment
    just vanished with no retry and no response.

    IdempotencyManager is a key -> PaymentResponse cache. A repeat request
    with a key already seen returns the cached result instead of charging
    the user twice -- this is what the original stub got wrong: it checked
    "not a duplicate" before submitting to the worker pool, but never
    checked the "is a duplicate" branch it also declared, and never records
    anything into the manager, so every repeat request would have
    reprocessed the payment.

    PaymentService wires a pluggable gateway, per-type payment methods, and
    the idempotency cache together. It hands gateway calls off to a
    goroutine over a channel (as the original stub's thread-pool intended)
    but -- unlike the original, which fired the task and threw the result
    away -- waits on the channel so CreatePayment can actually return the
    outcome to the caller.

Core Entities:
    PaymentRequest / PaymentResponse / PaymentType / PaymentStatus
    PaymentMethod (Card / UPI / NetBanking)
    PaymentGateway (RazorPay / Stripe) / RetryLogic
    IdempotencyManager
    PaymentService
================================================================================
*/

package main

import (
	"fmt"
	"strings"
	"sync"
)

type PaymentType int

const (
	Card PaymentType = iota
	UPI
	NetBanking
	Wallet
)

func (t PaymentType) String() string {
	switch t {
	case Card:
		return "CARD"
	case UPI:
		return "UPI"
	case NetBanking:
		return "NET_BANKING"
	case Wallet:
		return "WALLET"
	}
	return "UNKNOWN"
}

type PaymentStatus int

const (
	Success PaymentStatus = iota
	Failed
	Duplicate
)

func (s PaymentStatus) String() string {
	switch s {
	case Success:
		return "SUCCESS"
	case Failed:
		return "FAILED"
	case Duplicate:
		return "DUPLICATE"
	}
	return "UNKNOWN"
}

type PaymentRequest struct {
	IdempotencyKey string
	UserID         string
	Amount         float64
	Type           PaymentType
	Details        map[string]string
}

type PaymentResponse struct {
	PaymentID *string
	Status    PaymentStatus
	Message   string
}

func (r PaymentResponse) String() string {
	id := "null"
	if r.PaymentID != nil {
		id = *r.PaymentID
	}
	return fmt.Sprintf("PaymentResponse(payment_id=%s, status=%s, message=%s)", id, r.Status, r.Message)
}

type PaymentMethod interface {
	Validate(request PaymentRequest) bool
}

type CardPaymentMethod struct{}

func (CardPaymentMethod) Validate(request PaymentRequest) bool {
	cardNumber, ok := request.Details["cardNumber"]
	return ok && len(cardNumber) == 16
}

type UpiPaymentMethod struct{}

func (UpiPaymentMethod) Validate(request PaymentRequest) bool {
	upiID, ok := request.Details["upiId"]
	return ok && strings.Contains(upiID, "@")
}

type NetBankingPaymentMethod struct{}

func (NetBankingPaymentMethod) Validate(request PaymentRequest) bool {
	bankCode, ok := request.Details["bankCode"]
	return ok && bankCode != ""
}

type PaymentGateway interface {
	Process(request PaymentRequest) PaymentResponse
}

// RazorPayGateway simulates a gateway that fails the first attempt for a
// given idempotency key and succeeds on every attempt after that -- enough
// to exercise RetryLogic deterministically, without relying on real
// network flakiness.
type RazorPayGateway struct {
	mu       sync.Mutex
	attempts map[string]int
}

func NewRazorPayGateway() *RazorPayGateway {
	return &RazorPayGateway{attempts: make(map[string]int)}
}

func (g *RazorPayGateway) Process(request PaymentRequest) PaymentResponse {
	g.mu.Lock()
	g.attempts[request.IdempotencyKey]++
	attempt := g.attempts[request.IdempotencyKey]
	g.mu.Unlock()

	if attempt == 1 {
		return PaymentResponse{nil, Failed, "RazorPay: transient gateway error (attempt 1)"}
	}
	paymentID := "razorpay-" + request.IdempotencyKey
	return PaymentResponse{&paymentID, Success, fmt.Sprintf("RazorPay: payment captured (attempt %d)", attempt)}
}

// StripeGateway is a second, always-succeeding gateway -- demonstrates that
// PaymentService only depends on the PaymentGateway interface, so the
// processor behind it is pluggable.
type StripeGateway struct{}

func (StripeGateway) Process(request PaymentRequest) PaymentResponse {
	paymentID := "stripe-" + request.IdempotencyKey
	return PaymentResponse{&paymentID, Success, "Stripe: payment captured"}
}

type RetryLogic struct {
	gateway     PaymentGateway
	maxAttempts int
}

func NewRetryLogic(gateway PaymentGateway, maxAttempts int) *RetryLogic {
	return &RetryLogic{gateway: gateway, maxAttempts: maxAttempts}
}

func (r *RetryLogic) Process(request PaymentRequest) PaymentResponse {
	var lastResponse PaymentResponse
	for attempt := 1; attempt <= r.maxAttempts; attempt++ {
		lastResponse = r.gateway.Process(request)
		if lastResponse.Status == Success {
			return lastResponse
		}
	}
	return lastResponse
}

// IdempotencyManager maps key -> the response that was returned the first
// time that key was seen.
type IdempotencyManager struct {
	mu    sync.Mutex
	cache map[string]PaymentResponse
}

func NewIdempotencyManager() *IdempotencyManager {
	return &IdempotencyManager{cache: make(map[string]PaymentResponse)}
}

func (m *IdempotencyManager) GetCached(idempotencyKey string) (PaymentResponse, bool) {
	m.mu.Lock()
	defer m.mu.Unlock()
	response, ok := m.cache[idempotencyKey]
	return response, ok
}

func (m *IdempotencyManager) Record(idempotencyKey string, response PaymentResponse) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.cache[idempotencyKey] = response
}

type PaymentService struct {
	methods            map[PaymentType]PaymentMethod
	gateway            PaymentGateway
	idempotencyManager *IdempotencyManager
}

func NewPaymentService(gateway PaymentGateway) *PaymentService {
	return &PaymentService{
		methods: map[PaymentType]PaymentMethod{
			Card:       CardPaymentMethod{},
			UPI:        UpiPaymentMethod{},
			NetBanking: NetBankingPaymentMethod{},
		},
		gateway:            gateway,
		idempotencyManager: NewIdempotencyManager(),
	}
}

func (s *PaymentService) CreatePayment(request PaymentRequest) PaymentResponse {
	if cached, ok := s.idempotencyManager.GetCached(request.IdempotencyKey); ok {
		return PaymentResponse{cached.PaymentID, Duplicate,
			fmt.Sprintf("Duplicate request for idempotency key %s; returning cached result", request.IdempotencyKey)}
	}

	method, ok := s.methods[request.Type]
	if !ok || !method.Validate(request) {
		return PaymentResponse{nil, Failed, "Validation failed for " + request.Type.String()}
	}

	// Hand the gateway call off to a goroutine over a channel (mirroring a
	// worker-pool submission) and block for the result.
	resultCh := make(chan PaymentResponse, 1)
	go func() {
		resultCh <- s.gateway.Process(request)
	}()
	response := <-resultCh

	s.idempotencyManager.Record(request.IdempotencyKey, response)
	return response
}

func main() {
	service := NewPaymentService(NewRetryLogic(NewRazorPayGateway(), 3))

	fmt.Println("1) Valid card payment (gateway fails attempt #1, RetryLogic recovers it):")
	r1 := service.CreatePayment(PaymentRequest{"key-1", "user-1", 250.00, Card, map[string]string{"cardNumber": "4111111111111111"}})
	fmt.Println("  ", r1)

	fmt.Println("\n2) Same idempotency key resubmitted (must not charge again):")
	r2 := service.CreatePayment(PaymentRequest{"key-1", "user-1", 250.00, Card, map[string]string{"cardNumber": "4111111111111111"}})
	fmt.Println("  ", r2)

	fmt.Println("\n3) Invalid UPI id (missing '@'):")
	r3 := service.CreatePayment(PaymentRequest{"key-2", "user-2", 75.00, UPI, map[string]string{"upiId": "not-a-valid-upi-id"}})
	fmt.Println("  ", r3)

	fmt.Println("\n4) Valid UPI payment on a new idempotency key (also recovers via retry):")
	r4 := service.CreatePayment(PaymentRequest{"key-3", "user-2", 75.00, UPI, map[string]string{"upiId": "user2@upi"}})
	fmt.Println("  ", r4)

	fmt.Println("\n5) Swapping in Stripe for a net-banking payment (gateway is pluggable):")
	stripeService := NewPaymentService(NewRetryLogic(StripeGateway{}, 3))
	r5 := stripeService.CreatePayment(PaymentRequest{"key-4", "user-3", 500.00, NetBanking, map[string]string{"bankCode": "HDFC0001"}})
	fmt.Println("  ", r5)
}
